src/drivers/imx/sai.c

```diff
@@ -0,0 +1,343 @@
+	/* turn on (set to zero) stereo slot */
+	dai_update_bits(dai, REG_SAI_XMR(REG_TX_DIR),  REG_SAI_XMR_MASK,
+			~(BIT(0) | BIT(1)));
+
+	dai_update_bits(dai, REG_SAI_XCR2(REG_RX_DIR), mask_cr2, val_cr2);
+	dai_update_bits(dai, REG_SAI_XCR4(REG_RX_DIR), mask_cr4, val_cr4);
+	dai_update_bits(dai, REG_SAI_XCR5(REG_RX_DIR), mask_cr5, val_cr5);
+	/* turn on (set to zero) stereo slot */
+	dai_update_bits(dai, REG_SAI_XMR(REG_RX_DIR), REG_SAI_XMR_MASK,
+			~(BIT(0) | BIT(1)));
+
+	return 0;
+}
+
+static int sai_trigger(struct dai *dai, int cmd, int direction)
+{
+	trace_sai("SAI: sai_trigger");
+
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+		sai_start(dai, direction);
+		break;
+	case COMP_TRIGGER_STOP:
+		sai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_PAUSE:
+		sai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_RELEASE:
+		break;
+	case COMP_TRIGGER_SUSPEND:
+		break;
+	case COMP_TRIGGER_RESUME:
+		break;
+	default:
+		trace_sai_error("SAI: invalid trigger cmd %d", cmd);
+		break;
+	}
+	return 0;
+}
+
+static int sai_probe(struct dai *dai)
+{
+	trace_sai("SAI: sai_probe");
+
+	/* Software Reset for both Tx and Rx */
+	dai_update_bits(dai, REG_SAI_TCSR, REG_SAI_CSR_SR, REG_SAI_CSR_SR);
+	dai_update_bits(dai, REG_SAI_RCSR, REG_SAI_CSR_SR, REG_SAI_CSR_SR);
+
+	/* Clear SR bit to finish the reset */
+	dai_update_bits(dai, REG_SAI_TCSR, REG_SAI_CSR_SR, 0U);
+	dai_update_bits(dai, REG_SAI_RCSR, REG_SAI_CSR_SR, 0U);
+
+	/* Reset all Tx register values */
+	dai_write(dai, REG_SAI_TCR2, 0U);
+	dai_write(dai, REG_SAI_TCR3, 0U);
+	dai_write(dai, REG_SAI_TCR4, 0U);
+	dai_write(dai, REG_SAI_TCR5, 0U);
+	dai_write(dai, REG_SAI_TMR,  0U);
+
+	/* Reset all Rx register values */
+	dai_write(dai, REG_SAI_RCR2, 0U);
+	dai_write(dai, REG_SAI_RCR3, 0U);
+	dai_write(dai, REG_SAI_RCR4, 0U);
+	dai_write(dai, REG_SAI_RCR5, 0U);
+	dai_write(dai, REG_SAI_RMR,  0U);
+
+	sai_regs_dump(dai);
+	return 0;
+}
+
+static int sai_get_handshake(struct dai *dai, int direction, int stream_id)
+{
+	int handshake = dai->plat_data.fifo[direction].handshake;
+	int channel = EDMA_HS_GET_CHAN(handshake);
+	int irq = EDMA_HS_GET_IRQ(handshake);
+
+	return EDMA_HANDSHAKE(irq, channel);
+}
+
+static int sai_get_fifo(struct dai *dai, int direction, int stream_id)
+{
+	switch (direction) {
+	case DAI_DIR_PLAYBACK:
+	case DAI_DIR_CAPTURE:
+		return dai_fifo(dai, direction); // stream_id is unused
+	default:
+		trace_sai_error("sai_get_fifo(): Invalid direction");
+		return -EINVAL;
+	}
+}
+
+const struct dai_driver sai_driver = {
+	.type = SOF_DAI_IMX_SAI,
+	.dma_dev = DMA_DEV_SAI,
+	.ops = {
+		.trigger		= sai_trigger,
+		.set_config		= sai_set_config,
+		.pm_context_store	= sai_context_store,
+		.pm_context_restore	= sai_context_restore,
+		.probe			= sai_probe,
+		.get_handshake		= sai_get_handshake,
+		.get_fifo		= sai_get_fifo,
+	},
+};
+
+static void sai_regs_dump(struct dai *dai)
+{
+#define GET_SAI_REG(reg) dai_read(dai, REG_SAI_##reg)
```

Comment:
I think we should not add this function in the code. At least not in the initial submission. 

Also, I've opened a discussion about this on SOF community. Lets move the discussion thread there:

https://github.com/orgs/thesofproject/teams/sof-developers/discussions/24

Reply:
removed, it can maybe come back later if we want but under ifdef